Build: Transpile whitelisted packages in node_modules

blowery opened the pull request:
In the past, we've tried very hard to avoid transpiling code in node_modules. This was because most code shipped in node_modules was plain ES5 that worked in all of our target browsers, so transpiling would be extra work with no benefit. Modules were typically ES5 because they wanted to support Node 4, which was on par with IE11 for language features.

Fast forward to today, where Node 4 is no longer supported. Some modules (debug, d3-*, striptags, ...) are dropping support for Node 4 and adding ES6 features to the code they ship to NPM, like `let` and `for of`. This breaks IE11, so we have to transpile these packages to use them in IE11.

This PR proposes to transpile some things in node_modules, provided by a whitelist. This hasn't seemed to affect total compilation time very much, especially for incremental runs when the babel cache is hot. We also use a reduced babel config that only transpiles the set of plugins supported by `@babel/preset-env`, reducing the surface area of features we'll transpile. We're not really expecting anything outside of the `node@6` feature set, but using preset-env gives us a bit of future proofing.

The whitelist is path based, so it has no notion of a package's version. It would be a nice enhancement to only transpile packages that are past a certain version. `debug@4`, for instance, requires transpilation, but `debug@3` does not. We may have both versions in the tree as transitive dependencies.

**Original PR Description follows**
> In the past, we've tried very hard to avoid transpiling code in node_modules. This was because most code shipped in node_modules was plain ES5 that worked in all of our target browsers, so transpiling would be extra work with no benefit. Modules were typically ES5 because they wanted to support Node 4, which was on par with IE11 for language features.
> 
> Fast forward to today, were Node 4 is no longer supported. Some modules (debug, d3-*, striptags, ...) are dropping support for Node 4 and adding ES6 features to the code they ship to NPM, like `let` and `for of`. This breaks IE11, so we have to transpile these packages to use them in IE11.
> 
> This PR proposed to transpile _most_ things in node_modules. It provides a blacklist of modules which should be skipped, but most things will run through a basic transpilation. This hasn't seemed to affect total compilation time very much, especially for incremental runs where the babel cache is hot. We also use a reduced babel config that only transpiles the set of plugins supported by `@babel/preset-env`, reducing the surface area of features we'll transpile. We're not really expecting anything outside of the `node@6` feature set, but using preset-env gives us a bit of future proofing.
> 
> We may run into issues transpiling some packages, especially those that ship a minified version as the primary distribution. For those cases we provide a blacklist to exclude the package from transpilation. 
> 
> A couple upgrades have slipped in and IE11 is broken on some screens (thankfully not login or home) when modules like d3-array@2 are used.

blowery committed:
Try to transpile node_modules

This would let us use many of the newer NPM modules that are dropping support for ES6 and requiring Node 4, 6, 10.

blowery commented:
> An idea how we could auto-test for untranspiled code:

I think I'd rather just transpile it all if we can and avoid trying to detect what really needs it. Transpilation is pretty fast now and babel has a quality cache.

jsnajdr commented:
Interesting. TIL that CRA does that and that transpiling all `node_modules` is safe and reasonably fast.

blowery committed:
Split off the babel config for things in node_modules.
Teach that config to detect the module type.
Teach webpack about the new config.

jsnajdr commented:
> Do we need to transpile for tests too? Tests run in node and all this fancy syntax in node_modules should run on Node 10, so ... probably not?

We transpile tests with current Node as a target: https://github.com/Automattic/wp-calypso/blob/master/babel.config.js#L100-L107

Standard ES2018 should therefore transpile without any changes. The target is JSX and nonstandard extensions like class properties. Do any NPM modules ship that kind of syntax?

blowery commented:
> Do any NPM modules ship that kind of syntax?

Not that I've seen. This is mostly motivated by packages that are dropping old Node support and in the process picking up ES6 features, like `let`, `for of`, those sorts of things. I've not seen anyone shipping experimental syntax.

jsnajdr commented on babel.dependencies.config.js:
```diff
@@ -0,0 +1,31 @@
+				shippedProposals: true, // allows es7 features like Promise.prototype.finally
+			},
+		],
+		'@babel/react',
```
This transform shouldn't be needed. We're transpiling only ES6+ syntax to ES5, nothing else.

blowery replied:
```suggestion
```

jsnajdr commented on babel.dependencies.config.js:
```diff
@@ -0,0 +1,31 @@
+		'@babel/react',
+	],
+	plugins: [
+		'@babel/plugin-syntax-dynamic-import',
```
This transform is also not needed. We don't expect `import( ... ).then( ... )` statements in NPM packages, do we?

blowery replied:
I don't think so, but CRA has it.... It would be odd.

jsnajdr replied:
To CRA it was added by @Timer in https://github.com/facebook/create-react-app/pull/5047. The PR description doesn't mention that any published NPM package would use it.

It's a very minor issue that doesn't really matter at all. It's just my curiosity that's not 100% satisfied 😃

Timer replied:
> This transform is also not needed. We don't expect import( ... ).then( ... ) statements in NPM packages, do we?

We expect npm packages to use it in Create React App.

Timer replied:
Also, compiling `node_modules` is not without its pitfalls, it's very unsafe.
You should avoid compiling `node_modules if you can. 😄

blowery replied:
> You should avoid compiling `node_modules` if you can. 😄

Given the direction that many dependencies are going, dropping support for Node 4, I don't really think that's viable. At least not for IE11 and other downlevel browsers.

I'm not entirely sure what's unsafe about it though. @Timer can you share anything that CRA has run into?

Timer replied:
Some code just isn't safe to compile, or you hit edge cases when you double minify.
Mapbox GL was a big one that was broken, but many things that use web workers break IIRC.

The reasons for breaks are so nebulous I can't really quantify them down to a handful of reasons, sorry!

blowery replied:
@Automattic/calypso-sdk @jeffersonrabb could you give this PR a spin for the map block and see if we have problems?

jeffersonrabb replied:
Hi @blowery - I tried the PR out for Map and something is definitely broken. I compared this branch to `master` - on `master` all is well, but on this one the map tiles don't seem to be loading. Lots of errors in the console, although they're not exactly clear:

```blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1 Uncaught ReferenceError: _typeof is not defined
    at tr (blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1)
    at dr (blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1)
    at new bn (blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1)
    at new Cn (blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1)
    at blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1
    at blob:http://localhost/9821f18e-55d0-4fea-a0cb-16ec87408991:1```

blowery replied:
thanks @jeffersonrabb! i'll add mapbox to the blacklist.

blowery replied:
added in https://github.com/Automattic/wp-calypso/pull/30196/commits/34a434b381b17a4f66b1439b429e22a84e9cf262

blowery replied:
though that does make me revisit the choice to use a blacklist. Maybe we need a whitelist instead.

Timer replied:
I'd suggest keeping a list of "bad actors". If packages are intended to be used in the browser (even if they're Node packages) the publisher is responsible for compiling.
We always told users who tried to import non-ES5 that the publisher had to fix, and most of the time, they did.

Bad actors would just contain often-used packages that refuse to compile themselves.

Timer replied:
Insider tip: be sure to disable `transform-typeof-symbol`. This should fix mapbox and reduce some of the side effects of compiling `node_modules` if you go through with it.

blowery replied:
switch all this over to using a whitelist

jsnajdr commented on webpack.config.js:
```diff
@@ -211,9 +211,6 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
 		module: {
-			// avoids this warning:
-			// https://github.com/localForage/localForage/issues/577
-			noParse: /[\/\\]node_modules[\/\\]localforage[\/\\]dist[\/\\]localforage\.js$/,
```
The webpack warning that originally caused this issue doesn't seem to be in current versions any more.

jsnajdr commented on webpack.config.js:
```diff
@@ -236,6 +233,28 @@ function getWebpackConfig( { cssFilename, externalizeWordPressPackages = false }
+				{
+					test: /\.jsx?$/,
+					include: /node_modules\//,
+					exclude: /@babel(?:\/|\\{1,2})runtime/,
```
There are several libraries that we can trust to be ES5:
- `babel-runtime` -- the 6 version used still by many libraries: `react-virtualized`, `click-outside`, `wpcom` (!)
- `core-js`
- everything in `@wordpress/*`
- `lodash`
- `regenerate-unicode-properties` -- used by the `@babel/plugin-proposal-unicode-property-regex` transform
- `gridicons`
- `fbjs`
- `react` and related
- `wpcom`

jsnajdr replied:
One strange file that gets transpiled by Babel:
```
node_modules/css-loader/lib/css-base.js
```
I don't know why exactly, but it's probably the JS code that's added by `css-loader` and then removed again by the extract plugin?

blowery replied:
Looks like `exclude` can be a function instead of a regex. Perhaps that would make building and maintaining that list easier. https://webpack.js.org/configuration/module/#condition

blowery replied:
btw, `react` is not safe. They use Map and Set. :) https://reactjs.org/docs/javascript-environment-requirements.html

jsnajdr replied:
> btw, react is not safe. They use Map and Set.

That is safe for us. We import `@babel/polyfill` at the very beginning, and it polyfills the `Map` and `Set` globals, according to the browser requirements in `@babel/preset-env`.

It wouldn't be safe if we:
- used `useBuiltins: 'usage'` in `preset-env`. Babel needs to transpile a module to see that a polyfill is needed.
- used `corejs: true` in `transform-runtime`. Then Babel transpiles usages of the `Map` global to imports from `babel-runtime-corejs`, avoiding the need for global namespace pollution.

In both cases, Babel would need to see the React module to polyfill correctly.

blowery replied:
switched to a function based exclude in https://github.com/Automattic/wp-calypso/pull/30196/commits/40d46b81e3a71a1b87791a587d11882d6fd17803

also looked at trying to automate the set of includes with a function that parses each package.json and checks the `engines` property, but it looks like that won't fly because at least some of the packages that have dropped node 4 support have not updated `engines`.

jsnajdr reviewed (commented):
Build speed doesn't seem to be significantly affected. It turns out that the modules from `node_modules` are in minority. There are 8520 files transpiled, only 3540 of them coming from `node_modules`.

I gathered all this data by placing a `console.log` statement to the `babel-loader` source.

blowery committed:
use a function to determine excludes

blowery committed:
exclude mapbox-gl from transpilation

blowery committed:
remove @babel/react

blowery committed:
Disable babel transform-typeof-symbol

ockham commented:
I tend to agree with @Timer that it might make sense to err on the safe side when transpiling `node_modules` -- but I would love us to start doing it for those packages that actually publish their untranspiled code (so we can drop [some of these aliases](https://github.com/Automattic/wp-calypso/blob/40f010e7f0d891e69464ab849822e18b4fd707ca/webpack.config.js#L318-L322)).

Some prior art here: #19698. FWIW, the [heuristics I was using](https://github.com/Automattic/wp-calypso/pull/19698/files#diff-11e9f7f953edc64ba14b0cc350ae7b9dR74) was based on the presence of a `module` or `esnext` field in `package.json`. Granted, the former only means ESM (and not necessarily ES6(+)), and the latter is more of a [proposal](http://2ality.com/2017/06/pkg-esnext.html) at this point, but I thought there'd probably be a strong ESM-ES6(+) correlation (translating to 'untranspiled') anyway.

blowery commented:
@ockham I agree, we'll want to use a whitelist, not transpile everything. I'm not in love with using module / esnext as a heuristic, as it means we'll likely be transpiling more than we need to, but it might serve as a check. Maybe we also come up with a new thing that scans the source and yells if it sees things that need transpilation?

blowery committed:
switch to whitelist, fix up excludes, use new babel/env settings

blowery commented:
Switched this to use a whitelist in https://github.com/Automattic/wp-calypso/pull/30196/commits/aea84cfe3a2634efc042b44f7921d065e074070e

simison commented:
Noting that Mapbox dependency might go away soon (~this week) in Calypso after migrating the block's code from Calypso to Jetpack.

In Calypso
https://github.com/Automattic/wp-calypso/blob/e4862897aec6ccd7e3ef7120df19f20df216a07a/package.json#L143

In Jetpack
https://github.com/Automattic/jetpack/pull/11640/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R151

blowery commented:
@simison with the new whitelist, we're no longer trying to transpile mapbox-gl.

jsnajdr commented on webpack.config.js:
```diff
@@ -160,6 +193,27 @@ function getWebpackConfig( {
 				} ),
+				{
+					test: /\.jsx?$/,
+					include: _.memoize( shouldTranspileDependency ),
```
The `memoize` here is probably an insignificant optimization. It will maintain a map with several thousands entries, and webpack does so many checks on each filename so that this extra one is unlikely to make any impact.

jsnajdr commented on babel.dependencies.config.js:
```diff
@@ -0,0 +1,33 @@
+		],
+	],
+	// Exclude transforms that make all code slower, see https://github.com/facebook/create-react-app/pull/5278
+	exclude: [ 'transform-typeof-symbol' ],
```
Does the `exclude` need to be there in two copies?

jsnajdr reviewed (approved):
Posted two nits, looks good to :ship:

blowery committed:
remove duplicate exclude

ockham commented on webpack.config.js:
```diff
@@ -160,6 +193,27 @@ function getWebpackConfig( {
+							},
+						},
+					],
+				},
```
Could we use the `Transpile` shard from `@automattic/calypso-build` for this?

blowery replied:
possibly, i'll try it out

blowery replied:
done!

blowery committed:
drop memoize of included modules

blowery committed:
switch to using the transpile shard

blowery committed:
bump version on calypso-build

jsnajdr commented:
Thanks for fixing the nits! Looks good to me :+1:
